### PR TITLE
MP-ECDSA

### LINKED ISSUE #26: [CHECKLIST] MP-ECDSA Protocol
Code from the `dkg-gadget` crate of the `dkg-substrate` repo must be refactored inside the new gadget/protocol model provided by this repository to create a more robust, maintainable, and streamlined implementation.

Since the dkg has two types of protocols (keygen and signing), we can run two different protocols via `webb_gadget::run_protocol` selected twice, sharing some state that allows the signing to use the keygen's output public key. Thus, this task will include these two different phases.

**Phase I: Keygen Protocol**
- [x] #27
- [x] #28

**Phase II: Signing Protocol**
- [x] #29
- [x] #30

Ideally, there would be no need for an in-memory shared state between the `KeygenProtocol` and the `SigningProtocol`. Instead, we would use the off-chain storage as we currently do in the DKG to fetch keys for a specific session.

Since the function `webb_gadget::run_protocol` also requires a connection to a blockchain and a network implementation, we will also need to complete the following tasks:

**Phase III: Networking**
- [x] #31
- [x] #32
- [x] #39

Note: the client should be identical between the keygen and signing protocols. The network will be nearly identical, the only difference being that the underlying network stream will be multiplexed to split signing and keygen messages.

Next, the two protocols for keygen and signing will need to be put together as such:

```rust
let keygen_protocol_future = webb_gadget::run_protocol(keygen_gossip_network, keygen_protocol, blockchain_connection.clone());
let signing_protocol_future = webb_gadget::run_protocol(signing_gossip_network, signing_protocol, blockchain_connection);

tokio::select! {
    res0 = keygen_protocol_future => res0,
    res1 = signing_protocol_future => res1,
}
```

Thus, this step of tying together the two protocols and creating a program entrypoint with a main function will need to be encapsulated in an additional step:

**Phase IV: Application Binary**
- [ ] #33

Finally, everything must be tested:

**Phase V: Testing**
- [x] #34
- [ ] #35

### PR DESCRIPTION
**Pull request by tbraun96**
Closes #26

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -5,19 +5,28 @@ members = [
...
 ]
 
 [workspace.dependencies]
 gadget-core = { path = "./gadget-core" }
 webb-gadget = { path = "./webb-gadget" }
 zk-gadget = { path = "./zk-gadget" }
 zk-playground = { path = "./playground" }
+mp-ecdsa-protocol = { path = "./protocols/mp-ecdsa" }
+
+dkg-runtime-primitives = { git = "https://github.com/webb-tools/dkg-substrate" }
```
Comment (drewstone) [inline Cargo.toml:19]: Would prefer you don't bring in any of these crates @tbraun96. You should copy over only what needs to be brought over from there.

**Review by drewstone — CHANGES_REQUESTED** [review]

File: protocols/mp-ecdsa/src/client.rs [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+    B: Block,
+    BE: Backend<B>,
+    T: BlockchainEvents<B> + HeaderBackend<B> + ProvideRuntimeApi<B> + Send + Sync,
+    <T as ProvideRuntimeApi<B>>::Api:
+        DKGApi<B, AuthorityId, NumberFor<B>, MaxProposalLength, MaxAuthorities>,
+{
+}
+
+impl<B: Block, BE: Backend<B>, C: ClientWithApi<B, BE>> MpEcdsaClient<B, BE, C> {
+    pub async fn validator_set(
```
Comment (drewstone) [inline protocols/mp-ecdsa/src/client.rs:54]: This isn't exactly relevant anymore. The parties participating in the job are taken directly from the job.

Comment (drewstone) [reply protocols/mp-ecdsa/src/client.rs:54]: You can get the active validator set from the session API now or using `subxt`.

File: protocols/mp-ecdsa/src/client.rs [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+            }
+        };
+
+        self.logger
+            .trace(format!("🕸️  active validator set: {new:?}"));
+
+        new
+    }
+
+    pub async fn should_execute_new_keygen(
```
Comment (drewstone) [inline protocols/mp-ecdsa/src/client.rs:87]: This isn't relevant and will not be included in the runtime of the new system.

File: protocols/mp-ecdsa/src/client.rs [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+            .exec_client_function(move |client| {
+                client.runtime_api().should_execute_new_keygen(at).unwrap_or_default()
+            })
+            .await;
+
+        AnticipatedKeygenExecutionStatus { execute, force_execute }
+    }
+
+    /// Get the next DKG public key
+    pub async fn get_next_dkg_pub_key(
```
Comment (drewstone) [inline protocols/mp-ecdsa/src/client.rs:103]: There is no next public key anymore, there is just a Phase 1 Job that indicates completing a DKG w/ specified configuration parameters such as validator participants.

File: protocols/mp-ecdsa/src/client.rs [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+        header: &B::Header,
+    ) -> Option<(AuthoritySetId, Vec<u8>)> {
+        let at = header.hash();
+        self.exec_client_function(move |client| {
+            client.runtime_api().next_dkg_pub_key(at).unwrap_or_default()
+        })
+            .await
+    }
+
+    pub async fn dkg_pub_key_is_unset(&self, header: &B::Header) -> bool {
```
Comment (drewstone) [inline protocols/mp-ecdsa/src/client.rs:114]: This isn't something that will be on Tangle mainnet.

File: protocols/mp-ecdsa/src/client.rs [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+        self.get_dkg_pub_key(header).await.1.is_empty()
+    }
+
+    /// Get the active DKG public key
+    pub async fn get_dkg_pub_key(&self, header: &B::Header) -> (AuthoritySetId, Vec<u8>) {
```
Comment (drewstone) [inline protocols/mp-ecdsa/src/client.rs:119]: This isn't something that will be on Tangle mainnet.

File: protocols/mp-ecdsa/src/client.rs [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+        self.exec_client_function(move |client| {
+            client.runtime_api().dkg_pub_key(at).unwrap_or_default()
+        })
+            .await
+    }
+
+    /// Get the party index of our worker
+    ///
+    /// Returns `None` if we are not in the best authority set
+    pub async fn get_party_index(&self, header: &B::Header) -> Option<u16> {
```
Comment (drewstone) [inline protocols/mp-ecdsa/src/client.rs:130]: There will be no best authorities on Tangle Mainnet as is currently implemented by `dkg-substrate`. There is just the job and the participants of the job. The party index for a validator is their index in the job's participants.

File: protocols/mp-ecdsa/src/client.rs [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+            if elt.1 == public {
+                return Some(elt.0)
+            }
+        }
+
+        None
+    }
+
+    /// Get the best authorities for keygen
+    pub async fn get_best_authorities(&self, header: &B::Header) -> Vec<(u16, AuthorityId)> {
```
Comment (drewstone) [inline protocols/mp-ecdsa/src/client.rs:143]: There will be no best authorities on Tangle Mainnet as is currently implemented by dkg-substrate. There is just the job and the participants of the job. The party index for a validator is their index in the job's participants.

File: protocols/mp-ecdsa/src/client.rs [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+        self.exec_client_function(move |client| {
+            client.runtime_api().get_best_authorities(at).unwrap_or_default()
+        })
+            .await
+    }
+
+    /// Get the next party index of our worker for possible queued keygen
+    ///
+    /// Returns `None` if we are not in the next best authority set
+    pub async fn get_next_party_index(&self, header: &B::Header) -> Option<u16> {
```
Comment (drewstone) [inline protocols/mp-ecdsa/src/client.rs:154]: There is no next set of parties. There will be just a job and its participants.

File: protocols/mp-ecdsa/src/client.rs [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+            if elt.1 == public {
+                return Some(elt.0)
+            }
+        }
+
+        None
+    }
+
+    /// Get the next best authorities for keygen
+    pub async fn get_next_best_authorities(&self, header: &B::Header) -> Vec<(u16, AuthorityId)> {
```
Comment (drewstone) [inline protocols/mp-ecdsa/src/client.rs:167]: There is no next best authorities on Tangle mainnet. Just a job and its configuration parameters.

File: protocols/mp-ecdsa/src/client.rs [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+        let at = header.hash();
+        self.exec_client_function(move |client| {
+            client.runtime_api().get_next_best_authorities(at).unwrap_or_default()
+        })
+            .await
+    }
+
+    /// Get the signature threshold at a specific block
+    pub async fn get_signature_threshold(&self, header: &B::Header) -> u16 {
```
Comment (drewstone) [inline protocols/mp-ecdsa/src/client.rs:176]: There is no main signature threshold and the client API won't implement this as you have it here.

File: protocols/mp-ecdsa/src/client.rs [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+        let at = header.hash();
+        self.exec_client_function(move |client| {
+            client.runtime_api().signature_threshold(at).unwrap_or_default()
+        })
+            .await
+    }
+
+    /// Get the next signature threshold at a specific block
+    pub async fn get_next_signature_threshold(&self, header: &B::Header) -> u16 {
```
Comment (drewstone) [inline protocols/mp-ecdsa/src/client.rs:185]: There is no next signature threshold and the client API won't implement this as you have it here.

File: protocols/mp-ecdsa/src/constants.rs [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+    pub const MAX_KEYGEN_RETRIES: usize = 5;
+
+    /// How many blocks to keep the proposal hash in out local cache.
+    pub const PROPOSAL_HASH_LIFETIME: u32 = 10;
+}
+
+// ============= Signing Manager ======================= //
+
+pub mod signing_manager {
+    use dkg_primitives::types::SSID;
```
Comment (drewstone) [inline protocols/mp-ecdsa/src/constants.rs:25]: Please move any `dkg-primitives` into this repo and create a package for it. Some name ideas
- `tangle-mpc-primitives`
- `tangle-dkg-primitives`
open to other suggestions.

File: protocols/mp-ecdsa/src/keyring.rs [hunk 14]
```diff
@@ -0,0 +1,144 @@
+// Copyright 2022 Webb Technologies Inc.
```
Comment (drewstone) [inline protocols/mp-ecdsa/src/keyring.rs:1]: You can update these license headers to be 2022-2023

File: protocols/mp-ecdsa/src/keyring.rs [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+impl From<Keyring> for ecdsa::Pair {
+    fn from(k: Keyring) -> Self {
+        k.pair().into()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::Keyring;
+    use dkg_runtime_primitives::crypto;
```
Comment (drewstone) [inline protocols/mp-ecdsa/src/keyring.rs:81]: Avoid bringing in any `dkg-substrate` packages.

File: protocols/mp-ecdsa/src/keystore.rs [hunk 16]
```diff
@@ -0,0 +1,354 @@
...
+use crate::error;
+use crate::util::DebugLogger;
+
+/// A DKG specific keystore implemented as a `Newtype`. This is basically a
+/// wrapper around [`Keystore`] and allows to customize
+/// common cryptographic functionality.
+#[derive(Clone)]
+pub struct DKGKeystore(Option<KeystorePtr>, DebugLogger);
+
+impl DKGKeystore {
```
Comment (drewstone) [inline protocols/mp-ecdsa/src/keystore.rs:38]: If this is the MpEcdsa keystore we should name it such, I think getting more specific and semantically correct names would be beneficial here.

File: protocols/mp-ecdsa/src/protocols/managers/keygen.rs [hunk 17]
```diff
@@ -0,0 +1,554 @@
...
+                    header,
+                    anticipated_execution,
+                )
+                .await;
+        }
+
+        None
+    }
+
+    /// Check to see if we need to run a keygen for session 2, 3, 4, .., etc.
```
Comment (drewstone) [inline protocols/mp-ecdsa/src/protocols/managers/keygen.rs:292]: This isn't a thing anymore. If there's confusion here you need to clarify how we can assist.

File: protocols/mp-ecdsa/src/protocols/managers/keygen.rs [hunk 18]
```diff
@@ -0,0 +1,554 @@
...
+        None
+    }
+
+    async fn maybe_start_keygen_for_stage(
+        &self,
+        stage: KeygenRound,
+        header: &B::Header,
+        anticipated_execution_status: &AnticipatedKeygenExecutionStatus,
+    ) -> Option<Job> {
+        let authority_set = if let Some((active, queued)) = self.client.validator_set(header).await {
```
Comment (drewstone) [inline protocols/mp-ecdsa/src/protocols/managers/keygen.rs:347]: Not how it'll go. This is leading me to believe you're not processing everything about the new architecture properly given all our previous discussions.

**Review by drewstone — CHANGES_REQUESTED** [review]
Some scaffold looks OK, but generally doesn't seem like the approach you've taken is correct given the new architecture. There is no base DKG anymore and it does not rotate on sessions. The client API from `dkg-substrate` is not being included in Tangle mainnet at all. There will be no best authorities, next best authorities, signature thresholds, etc. Everything should be tied to a job. After receiving job details we should execute keygen or signing protocols. CC @1xstj

File: protocols/mp-ecdsa/src/protocols/sign.rs [hunk 19]
```diff
@@ -0,0 +1,336 @@
...
+            if participants.contains(&self.account_id) {
+                let task_id = job.job_id.to_be_bytes();
+                let task_id = keccak_256(&task_id);
+                let session_id = 0; // We are not interested in sessions for the ECDSA protocol
+                let retry_id = 0; // TODO: query the job manager for the retry id
+
+                let key = job.key.expect("Should exist for stage 2 signing");
+                // TODO for keygen: use bincode2 for serialization
+                let deserialized_key: LocalKey<Secp256k1> =
+                    bincode2::deserialize(&key).expect("Should deserialize");
```
Comment (shekohex) [inline protocols/mp-ecdsa/src/protocols/sign.rs:90]: This `key` here in the job.key is not the `LocalKey`, Thomas. It is the Public Key of the DKG. The "LocalKey" is, by definition, Local to you, the current validator and never shared, please treat that as a private material. It should only exist in the validator database, encrypted.

File: protocols/mp-ecdsa/src/protocols/keygen.rs [hunk 20]
```diff
@@ -0,0 +1,202 @@
...
+                let local_key = round_based::AsyncProtocol::new(
+                    state_machine_wrapper,
+                    rx_async_proto,
+                    tx_to_outbound,
+                )
+                .set_watcher(StderrWatcher)
+                .run()
+                .await?;
+
+                key_store.set(additional_params.job_id, local_key).await?;
```
Comment (shekohex) [inline protocols/mp-ecdsa/src/protocols/keygen.rs:183]: This LocalKey, you should save the key by the `phase_one.job_id` this one you should use above (see my previous comment)

File: protocols/mp-ecdsa/src/util.rs [hunk 21]
```diff
@@ -26,3 +29,33 @@ impl DebugLogger {
...
+
+    for _ in 0..(32 - s.len()) {
+        sig_vec.extend([0]);
+    }
+    sig_vec.extend_from_slice(&s);
+
+    sig_vec.extend([v]);
+
+    if 65 != sig_vec.len() {
+        log::warn!(target: "dkg_gadget", "🕸️  Invalid signature len: {}, expected 65", sig_vec.len());
```
Comment (drewstone) [inline protocols/mp-ecdsa/src/util.rs:53]: Don't forget to remove dkg_gadget references.

**Comment by shekohex** [comment]
Looks good. @tbraun96 lets rename webb-gadget to gadget-common and move the client.rs implementation there. Which will be reusable between all gadgets.

File: gadget-common/Cargo.toml [hunk 22]
```diff
@@ -19,4 +19,8 @@ sc-client-api = { workspace = true }
 sp-core = { workspace = true }
 sp-runtime = { workspace = true }
 hex = { workspace = true }
-bincode2 = { workspace = true }
\ No newline at end of file
+bincode2 = { workspace = true }
+
+sp-api = { workspace = true, default-features = false, features = ["std"] }
+tangle-primitives = { workspace = true, default-features = false, features = ["std"] }
+pallet-jobs-rpc-runtime-api = { workspace = true, default-features = false, features = ["std"] }
```
Comment (shekohex) [inline gadget-common/Cargo.toml:26]: Lets make `std` feature in that crate, and only enable `std` for other packages/crates only if we enabled `std`

**Review by shekohex — CHANGES_REQUESTED** [review]
Looks good! Just one comment about std feature, to avoid any conflicts when compiling tangle node/runtime.

**Review by shekohex — APPROVED** [review]
LGTM!

**Pull request merged**